zcash_primitives/src/transaction/components/sapling.rs

```diff
@@ -40,49 +117,84 @@ impl std::fmt::Debug for SpendDescription {
     }
 }
 
-impl SpendDescription {
-    pub fn read<R: Read>(mut reader: &mut R) -> io::Result<Self> {
-        // Consensus rules (§4.4):
-        // - Canonical encoding is enforced here.
-        // - "Not small order" is enforced in SaplingVerificationContext::check_spend()
-        //   (located in zcash_proofs::sapling::verifier).
-        let cv = {
-            let mut bytes = [0u8; 32];
-            reader.read_exact(&mut bytes)?;
-            let cv = jubjub::ExtendedPoint::from_bytes(&bytes);
-            if cv.is_none().into() {
-                return Err(io::Error::new(io::ErrorKind::InvalidInput, "invalid cv"));
-            }
-            cv.unwrap()
-        };
+/// Consensus rules (§4.4) & (§4.5):
+/// - Canonical encoding is enforced here.
+/// - "Not small order" is enforced in SaplingVerificationContext::(check_spend()/check_output())
+///   (located in zcash_proofs::sapling::verifier).
+pub fn read_point<R: Read>(mut reader: R, field: &str) -> io::Result<jubjub::ExtendedPoint> {
+    let mut bytes = [0u8; 32];
+    reader.read_exact(&mut bytes)?;
+    let point = jubjub::ExtendedPoint::from_bytes(&bytes);
+
+    if point.is_none().into() {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "invalid ".to_owned() + field,
+        ))
+    } else {
+        Ok(point.unwrap())
+    }
+}
 
-        // Consensus rule (§7.3): Canonical encoding is enforced here
-        let anchor = {
-            let mut f = [0u8; 32];
-            reader.read_exact(&mut f)?;
-            bls12_381::Scalar::from_repr(f)
-                .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "anchor not in field"))?
-        };
+/// Consensus rules (§7.3) & (§7.4):
+/// - Canonical encoding is enforced here
+pub fn read_scalar<R: Read>(mut reader: R, field: &str) -> io::Result<bls12_381::Scalar> {
```

Comment:
It's confusing to call this `read_scalar`. It's reading a BLS12-381 scalar, which is a Jubjub base field element, and it's the latter interpretation we want (as anchors are extracted from Jubjub curve point coordinates). Meanwhile, we _do_ read Jubjub scalars in other places.

At some point we should add a `type Base = bls12_381::Scalar` to `jubjub`, but in the meantime:
```suggestion
pub fn read_base<R: Read>(mut reader: R, field: &str) -> io::Result<bls12_381::Scalar> {
```